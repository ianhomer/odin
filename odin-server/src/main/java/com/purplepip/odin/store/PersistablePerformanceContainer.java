/*
 * Copyright (c) 2017 the original author or authors. All Rights Reserved
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.purplepip.odin.store;

import com.purplepip.odin.creation.channel.Channel;
import com.purplepip.odin.creation.layer.MutableLayer;
import com.purplepip.odin.creation.sequence.SequenceConfiguration;
import com.purplepip.odin.performance.PerformanceContainer;
import com.purplepip.odin.server.rest.repositories.ChannelRepository;
import com.purplepip.odin.server.rest.repositories.LayerRepository;
import com.purplepip.odin.server.rest.repositories.SequenceRepository;
import com.purplepip.odin.store.domain.PersistableChannel;
import com.purplepip.odin.store.domain.PersistableLayer;
import com.purplepip.odin.store.domain.PersistableSequence;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
@Slf4j
public class PersistablePerformanceContainer extends PerformanceContainer {
  @Autowired
  private LayerRepository layerRepository;

  @Autowired
  private ChannelRepository channelRepository;

  @Autowired
  private SequenceRepository sequenceRepository;

  @Override
  public void addChannel(Channel channel) {
    if (channel instanceof PersistableChannel) {
      ((PersistableChannel) channel).setPerformance(getPerformance());
      channelRepository.save((PersistableChannel) channel);
    }
    super.addChannel(channel);
  }

  @Override
  public PersistablePerformanceContainer addSequence(SequenceConfiguration sequence) {
    LOG.debug("Adding sequence : {}", sequence);
    if (sequence instanceof PersistableSequence) {
      PersistableSequence persistableSequence = ((PersistableSequence) sequence);
      LOG.debug("Saving persistable sequence : {}", sequence);
      persistableSequence.setPerformance(getPerformance());
      sequenceRepository.save((PersistableSequence) sequence);
    }
    super.addSequence(sequence);
    return this;
  }

  @Override
  public PersistablePerformanceContainer addLayer(MutableLayer layer) {
    if (layer instanceof PersistableLayer) {
      ((PersistableLayer) layer).setPerformance(getPerformance());
      layerRepository.save((PersistableLayer) layer);
    }
    super.addLayer(layer);
    return this;
  }
  // TODO : Create addTrigger implementation
}
