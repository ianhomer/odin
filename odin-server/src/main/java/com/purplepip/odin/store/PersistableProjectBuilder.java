/*
 * Copyright (c) 2017 the original author or authors. All Rights Reserved
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.purplepip.odin.store;

import com.purplepip.odin.composition.MutableSequenceConfiguration;
import com.purplepip.odin.composition.tick.Tick;
import com.purplepip.odin.composition.triggers.MutableTriggerConfiguration;
import com.purplepip.odin.math.Rational;
import com.purplepip.odin.math.Real;
import com.purplepip.odin.music.notes.Note;
import com.purplepip.odin.project.ProjectContainer;
import com.purplepip.odin.sequencer.BeanyProjectBuilder;
import com.purplepip.odin.sequencer.Channel;
import com.purplepip.odin.store.domain.PersistableChannel;
import com.purplepip.odin.store.domain.PersistableLayer;
import com.purplepip.odin.store.domain.PersistableNote;
import com.purplepip.odin.store.domain.PersistableSequence;
import com.purplepip.odin.store.domain.PersistableTick;
import com.purplepip.odin.store.domain.PersistableTrigger;

/**
 * Sequence builder that updates the project with persistable domain objects.
 */
public class PersistableProjectBuilder extends BeanyProjectBuilder {
  public PersistableProjectBuilder(ProjectContainer projectContainer) {
    super(projectContainer);
  }

  @Override
  protected PersistableLayer createLayer() {
    return new PersistableLayer();
  }

  @Override
  protected MutableSequenceConfiguration createSequence() {
    return new PersistableSequence();
  }

  /**
   * Create a persistable note.
   *
   * @return note
   */
  @Override
  protected Note createNote(int number, int velocity, Real duration) {
    PersistableNote persistableNote = new PersistableNote();
    if (duration instanceof Rational) {
      persistableNote.setNumerator(((Rational) duration).getNumerator());
      persistableNote.setDenominator(((Rational) duration).getDenominator());
    } else {
      // A good enough approximation, given that we'll mostly be dealing with rationals.
      persistableNote.setNumerator(duration.floor());
      persistableNote.setDenominator(1);
    }
    persistableNote.setVelocity(velocity);
    persistableNote.setNumber(number);
    persistableNote.afterLoad();
    return persistableNote;
  }

  @Override
  protected Channel createChannel() {
    return new PersistableChannel();
  }

  @Override
  protected MutableTriggerConfiguration createTrigger() {
    return new PersistableTrigger();
  }

  @Override
  protected Tick createTick(Tick tick) {
    return new PersistableTick(tick);
  }
}
