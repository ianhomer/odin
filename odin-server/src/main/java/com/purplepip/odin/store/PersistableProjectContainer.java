/*
 * Copyright (c) 2017 the original author or authors. All Rights Reserved
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.purplepip.odin.store;

import com.purplepip.odin.creation.channel.Channel;
import com.purplepip.odin.creation.layer.MutableLayer;
import com.purplepip.odin.creation.sequence.SequenceConfiguration;
import com.purplepip.odin.project.ProjectContainer;
import com.purplepip.odin.server.rest.repositories.ChannelRepository;
import com.purplepip.odin.server.rest.repositories.LayerRepository;
import com.purplepip.odin.server.rest.repositories.SequenceRepository;
import com.purplepip.odin.store.domain.PersistableChannel;
import com.purplepip.odin.store.domain.PersistableLayer;
import com.purplepip.odin.store.domain.PersistableSequence;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class PersistableProjectContainer extends ProjectContainer {
  @Autowired
  private LayerRepository layerRepository;

  @Autowired
  private ChannelRepository channelRepository;

  @Autowired
  private SequenceRepository sequenceRepository;

  @Override
  public void addChannel(Channel channel) {
    if (channel instanceof PersistableChannel) {
      channel.setProject(getProject());
      channelRepository.save((PersistableChannel) channel);
    }
    super.addChannel(channel);
  }

  @Override
  public PersistableProjectContainer addSequence(SequenceConfiguration sequence) {
    if (sequence instanceof PersistableSequence) {
      sequence.setProject(getProject());
      sequenceRepository.save((PersistableSequence) sequence);
    }
    super.addSequence(sequence);
    return this;
  }

  @Override
  public PersistableProjectContainer addLayer(MutableLayer layer) {
    if (layer instanceof PersistableLayer) {
      layer.setProject(getProject());
      layerRepository.save((PersistableLayer) layer);
    }
    super.addLayer(layer);
    return this;
  }
}
