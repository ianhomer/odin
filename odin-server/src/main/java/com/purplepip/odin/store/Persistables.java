/*
 * Copyright (c) 2017 Ian Homer. All Rights Reserved
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.purplepip.odin.store;

import com.purplepip.odin.store.domain.PersistableThing;

/**
 * Persistable utility functions.
 */
public final class Persistables {
  /**
   * Sometimes toSting() method on an entity is not safe, since the entity might have lazy
   * collections that have not been initialised.  An example of this situation is trying to
   * log the entity toString() after it has been deleted from the store.
   *
   * @param object object to output as string.
   * @return string representation
   */
  public static String toThingString(Object object) {
    String string;
    if (object instanceof PersistableThing) {
      /*
       * Just output the thing string which know will be safe since it does not have any
       * lazy properties.
       */
      string = ((PersistableThing) object).toThingString();
    } else {
      /*
       * If it's not a persistable thing, just call toString as normal.
       */
      string = object.toString();
    }
    return string;
  }

  private Persistables() {
  }
}
