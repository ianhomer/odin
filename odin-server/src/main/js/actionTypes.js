export const ADD_CHANNEL = 'ADD_CHANNEL'