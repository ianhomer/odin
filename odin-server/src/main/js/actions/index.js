'use strict'

export const ADD_CHANNEL = 'ADD_CHANNEL'
export const addChannel = (channel) => ({
  type: ADD_CHANNEL,
  channel
})

export const DELETE_CHANNEL = 'DELETE_CHANNEL'
export const deleteChannel = (channel) => ({
  type: DELETE_CHANNEL,
  channel
})

export const LOAD_ENTITIES_REQUESTED = 'LOAD_ENTITIES_REQUESTED'
export function loadEntitiesRequested(path, schema) {
  return {
    type: LOAD_ENTITIES_REQUESTED,
    path,
    schema: schema
  }
}

export const LOAD_ENTITIES_SUCCEEDED = 'LOAD_ENTITIES_SUCCEEDED'
export const LOAD_ENTITIES_FAILED = 'LOAD_ENTITIES_FAILED'

export const RECEIVE_ENTITIES = 'RECEIVE_ENTITIES'
export function receiveEntities(path, entities) {
  return {
    type: RECEIVE_ENTITIES,
    path,
    entities: entities,
    receivedAt: Date.now()
  }
}

export const CREATE_ENTITY_REQUESTED = 'CREATE_ENTITY_REQUESTED'
export function createEntityRequested(entity, path) {
  return {
    type: CREATE_ENTITY_REQUESTED,
    path,
    entity : entity
  }
}

export const CREATE_ENTITY_SUCCEEDED = 'CREATE_ENTITY_SUCCEEDED'
export const CREATE_ENTITY_FAILED = 'CREATE_ENTITY_FAILED'
