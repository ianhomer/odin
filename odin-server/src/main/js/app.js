'use strict';

const React = require('react');
const ReactDOM = require('react-dom');
const client = require('./client');

const ChannelList = require('./components/channelList')
const PatternList = require('./components/patternList')
const ProjectList = require('./components/projectList')
const SequenceList = require('./components/sequenceList')
const Trace = require('./components/trace')

class App extends React.Component {

	constructor(props) {
		super(props);
		this.state = {
		  sequences: [], projects: [],  project: null, pageSize: 10,
		};
	}

	componentDidMount() {
		client({method: 'GET', path: '/api/projects'}).done(response => {
		  var projects = response.entity._embedded.projects;
			this.setState({projects: projects, project: projects[0]});
		});
	}

	render() {
    // TODO : Switch to generic sequence list to configure more than just patterns
		return (
		  <div>
        <Trace scope="app"/>
		    {this.state.entities &&
		      <div className="warn">WARNING : Entities store in app state {JSON.stringify(this.state.entities)}</div>
		    }
        <ProjectList projects={this.state.projects}/>
        {this.state.project &&
          <ChannelList project={this.state.project}/>
        }
        {this.state.project &&
          <PatternList project={this.state.project}/>
        }
			</div>
		)
	}
}

ReactDOM.render(
	<App />,
	document.getElementById('react')
)
