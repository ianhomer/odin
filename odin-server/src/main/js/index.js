// Copyright (c) 2017 the original author or authors. All Rights Reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react'
import { render } from 'react-dom'

import thunkMiddleware from 'redux-thunk'
import { createLogger } from 'redux-logger'
import { applyMiddleware, combineReducers, createStore } from 'redux'
import 'babel-polyfill'
import createSagaMiddleware from 'redux-saga'

import App from './containers/app'
import reducers from './reducers'
import { Flux } from './legacy/flux'
import { Backend } from './backend'
import { loadSchemaActions } from './actions'

// Create the backend integration layer
const backend = new Backend()

// Create saga middleware for handling of calls to backend.  Note that the backend is
// passed in as context for the saga middleware so that we can inject the backend API calls in.
// Injecting backend API calls in is required for switching in different backends as required
// for test cases which need a mock backend.
const sagaMiddleware = createSagaMiddleware({context: {backend: backend}})

// Create the application store
const store = createStore(
  combineReducers(reducers),
  applyMiddleware(
    thunkMiddleware,
    sagaMiddleware,
    createLogger({collapsed: true})
  )
)

// Hook in back end API calls
sagaMiddleware.run(backend.saga)

for (var i = 0; i < loadSchemaActions.length ; i++) {
  store.dispatch(loadSchemaActions[i])
}

// Render application
render(
  <App store={store} flux={new Flux()}/>,
  document.getElementById('root')
)