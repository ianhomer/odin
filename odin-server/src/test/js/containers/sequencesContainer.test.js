import React from 'react'
import SequencesContainer from 'odin/containers/sequencesContainer.js'
import { LOAD_PROJECT_SCHEMA_SUCCEEDED, LOAD_PROFILE_SCHEMA_SUCCEEDED,
  loadSchemaActions } from 'odin/actions/index.js'
import { shallow } from 'enzyme'

import { dispatchAndExpect } from '../utils/dispatchAndExpect'
import { mockFlux, testProject, testSchema } from '../testData.js'
import store from '../store'

test('Sequences container renders OK', done => {
  const component =
    <SequencesContainer store={store} schema={testSchema} project={testProject} flux={mockFlux}/>
  expect(shallow(component).dive()).toMatchSnapshot()

  dispatchAndExpect(store, done,
    [LOAD_PROJECT_SCHEMA_SUCCEEDED, LOAD_PROFILE_SCHEMA_SUCCEEDED, LOAD_PROFILE_SCHEMA_SUCCEEDED],
    () => loadSchemaActions,
    () => {
      expect(shallow(component).dive()).toMatchSnapshot()
    }
  )
})