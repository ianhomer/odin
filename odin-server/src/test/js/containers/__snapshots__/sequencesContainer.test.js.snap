// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequences container renders OK 1`] = `
<div>
  <SequenceList
    entities={Array []}
    flux={
      MockFlux {
        "client": [Function],
        "loadFromServer": [Function],
        "loadSchema": [Function],
        "onCreate": [Function],
        "onDelete": [Function],
        "onPatch": [Function],
        "onUpdate": [Function],
      }
    }
    onCreate={[Function]}
    onDelete={[Function]}
    path="sequence"
    project={
      Object {
        "_links": Object {
          "channels": Object {
            "href": "http://localhost:8080/api/project/1/channels",
          },
          "layers": Object {
            "href": "http://localhost:8080/api/project/1/layers",
          },
          "project": Object {
            "href": "http://localhost:8080/api/project/1",
          },
          "self": Object {
            "href": "http://localhost:8080/api/project/1",
          },
          "sequences": Object {
            "href": "http://localhost:8080/api/project/1/sequences",
          },
        },
        "name": "defaultProject",
      }
    }
    schema={
      Schema {
        "revision": undefined,
      }
    }
  />
</div>
`;

exports[`Sequences container renders OK 2`] = `
<div>
  <SequenceList
    entities={Array []}
    flux={
      MockFlux {
        "client": [Function],
        "loadFromServer": [Function],
        "loadSchema": [Function],
        "onCreate": [Function],
        "onDelete": [Function],
        "onPatch": [Function],
        "onUpdate": [Function],
      }
    }
    onCreate={[Function]}
    onDelete={[Function]}
    path="sequence"
    project={
      Object {
        "_links": Object {
          "channels": Object {
            "href": "http://localhost:8080/api/project/1/channels",
          },
          "layers": Object {
            "href": "http://localhost:8080/api/project/1/layers",
          },
          "project": Object {
            "href": "http://localhost:8080/api/project/1",
          },
          "self": Object {
            "href": "http://localhost:8080/api/project/1",
          },
          "sequences": Object {
            "href": "http://localhost:8080/api/project/1/sequences",
          },
        },
        "name": "defaultProject",
      }
    }
    schema={
      Schema {
        "revision": undefined,
      }
    }
  />
</div>
`;
